File: 5주차/구민지/네트워크.py
from collections import deque

def bfs(start):
    global graph,visited
    if visited[start]:
        return 0
    cnt=0
    q=deque([start])
    while q:
        n=q.popleft()
        if visited[n]:
            continue
        visited[n]=True
        cnt+=1
        for i in graph[n]:
            if not visited[i]:
                q.append(i)
    return cnt
        
    
def solution(n, computers):
    answer = 0
    global graph,visited
    graph=[[] for _ in range(n+1)]
    visited=[False]*n
    for i in range(n):
        for j in range(n):
            if i==j:
                continue
            if computers[i][j]:
                graph[i].append(j)

    for i in range(n):
        if bfs(i):
            answer+=1
    return answer

File: 5주차/김용상/네트워크.java
import java.util.*;
class Solution {
    private boolean [] visited ;
    private int cnt = 0;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        visited = new boolean[n+1];
        Deque<Integer>dq = new ArrayDeque<>();
        List<Integer>[]graph = new ArrayList[n+1];
        for (int i=0;i<n;i++){
            graph[i] = new ArrayList<>();
        }
        for (int i=0;i<computers.length;i++){
            for (int j=0;j<computers[i].length;j++){
                if(computers[i][j]==1&&i!=j){
                    graph[i].add(j);
                    graph[j].add(i);
                }
            }
        }
        for (int i=0;i<n;i++){
            if (visited[i]){
                continue;
            }
            cnt++;
            dq.add(i);
            
            while(!dq.isEmpty()){
                int poll = dq.poll();
                if (visited[poll]){
                    continue;
                }
                visited[poll] = true;
                for (int k: graph[poll]){
                    dq.add(k);
                }
            }
        }
        
        
        
        return cnt;
    }
}

